libsolidity/lsp/LanguageServer.cpp

```diff
@@ -400,3 +426,89 @@ void LanguageServer::handleTextDocumentDidClose(MessageID _id, Json::Value const
 
 	compileAndUpdateDiagnostics();
 }
+
+ASTNode const* LanguageServer::requestASTNode(std::string const& _sourceUnitName, LineColumn const& _filePos)
+{
+	if (m_compilerStack.state() < CompilerStack::AnalysisPerformed)
+		return nullptr;
+
+	if (!m_fileRepository.sourceUnits().count(_sourceUnitName))
+		return nullptr;
+
+	optional<int> sourcePos = m_compilerStack.charStream(_sourceUnitName)
+		.translateLineColumnToPosition(_filePos);
+	if (!sourcePos.has_value())
+		return nullptr;
+
+	return locateInnermostASTNode(*sourcePos, m_compilerStack.ast(_sourceUnitName));
+}
+
+void LanguageServer::handleGotoDefinition(MessageID _id, Json::Value const& _args)
+{
+	string const uri = _args["textDocument"]["uri"].asString();
+	string const sourceUnitName = m_fileRepository.clientPathToSourceUnitName(uri);
+	if (!m_fileRepository.sourceUnits().count(sourceUnitName))
+	{
+		m_client.error(_id, ErrorCode::RequestFailed, "Unknown file: " + uri);
+		return;
+	}
+
+	auto const lineColumn = parseLineColumn(_args["position"]);
+	if (!lineColumn)
+	{
+		m_client.error(
+			_id,
+			ErrorCode::RequestFailed,
+			fmt::format(
+				"Unknown position {line}:{column} in file: {file}",
+				fmt::arg("line", lineColumn.value().line),
+				fmt::arg("column", lineColumn.value().column),
+				fmt::arg("file", sourceUnitName)
+			)
+		);
+		return;
+	}
+
+	ASTNode const* sourceNode = requestASTNode(sourceUnitName, lineColumn.value());
+	vector<SourceLocation> locations;
+	if (auto const* identifier = dynamic_cast<Identifier const*>(sourceNode))
```

Comment:
If you first check if it is an Expression you can group Identifier, IdentifierPath and MemberAccess, work with Declarations instead of SourceLocations and reduce quite some code duplicaiton.

Reply:
Maybe I misunderstood that, but `IdentifierPath` is not a base of `Expression`, but I tried to group the others however in front.